Hmusum/wait for config generation before restarting [run-systemtest]

hmusum opened the pull request.

Commit:
Add converged field to ServiceListResponse

Commit:
Add feature flag for checking config convergence before restart

Commit:
Move request url to http response

Commit:
Check config convergence before restarting services

Controlled by feature flag

jonmv commented on configserver/src/main/java/com/yahoo/vespa/config/server/ApplicationRepository.java:
```diff
@@ -218,6 +218,7 @@ private ApplicationRepository(TenantRepository tenantRepository,
         private Metric metric = new NullMetric();
         private SecretStoreValidator secretStoreValidator = new SecretStoreValidator(new SecretStoreProvider().get());
         private FlagSource flagSource = new InMemoryFlagSource();
+        private ConfigConvergenceChecker configConvergenceChecker = new ConfigConvergenceChecker();
```
This component is injectable, probably because it holds an executor. Perhaps use this?

On a side note, that executor isn't closed (no `deconstruct()` method), which should also be fixed.

hmusum replied:
Will look into it

jonmv commented on configserver/src/main/java/com/yahoo/vespa/config/server/deploy/Deployment.java:
```diff
@@ -149,6 +156,9 @@ private void restartServices(ApplicationId applicationId) {
 
         if ( ! restartActions.isEmpty()) {
+
+            waitForConfigToConverge(applicationId);
```
Who does the waiting here? I guess it could take a long time.

hmusum replied:
Yes, will do the same as in `waitForResources()`

jonmv commented on configserver/src/main/java/com/yahoo/vespa/config/server/deploy/Deployment.java:
```diff
@@ -165,6 +175,31 @@ private void restartServices(ApplicationId applicationId) {
+            }
+        } while (clock.instant().isBefore(end));
+
+        throw new RuntimeException("Config has not converged");
```
I suppose this'll ultimately lead to a failed deployment, not retried by the controller during deployment jobs. Should we retry? If so, we need some way to communicate this.

hmusum replied:
Yes, will look into it

jonmv reviewed (commented):
Some comments, otherwise LGTM!

Commit:
Uae timeout budget and improve some logging

Commit:
Add deconstruct() to ConfigConvergenceChecker

Commit:
Add note about when builder can be used

Commit:
Add TODO on how to handle restartOnDeploy for container clusters

Commit:
Use ConfigNotConvergedException, which will return 409

Return 409, which is also done for creatificate not ready and
parent host not ready when config has not converged before restarting

hmusum commented:
@jonmv please take a look again

jonmv reviewed (approved):
LGTM. I believe a mapping from this error to something that is retried is needed in the controller code if this should allow deployments to be retried there, but that can be added later.
